beluga_amcl/include/beluga_amcl/ndt_amcl_node_3d.hpp

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022-2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_AMCL_NDT_AMCL_NODE_3D_HPP
+#define BELUGA_AMCL_NDT_AMCL_NODE_3D_HPP
+
+#include <tf2_ros/buffer.h>
+#include <tf2_ros/message_filter.h>
+#include <tf2_ros/transform_broadcaster.h>
+#include <tf2_ros/transform_listener.h>
+
+#include <beluga/algorithm/amcl_core.hpp>
+#include <beluga/beluga.hpp>
+
+#include <execution>
+#include <functional>
+#include <optional>
+#include <sophus/types.hpp>
+#include <tuple>
+#include <variant>
+
+#include <geometry_msgs/msg/pose_array.hpp>
+#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
+#include <rclcpp_lifecycle/lifecycle_node.hpp>
+#include <sensor_msgs/msg/laser_scan.hpp>
+#include <sensor_msgs/msg/point_cloud2.hpp>
+#include <std_srvs/srv/empty.hpp>
+
+#include <beluga_ros/laser_scan.hpp>
+#include "beluga_amcl/ros2_common.hpp"
+
+#include <message_filters/subscriber.h>
+
+/**
+ * \file
+ * \brief ROS 2 integration of the 3D NDT-AMCL algorithm.
+ */
+
+namespace beluga_amcl {
+
+/// Underlying map representation for the NDT sensor model.
+using NDTMapRepresentation =
+    beluga::SparseValueGrid3<std::unordered_map<Eigen::Vector3i, beluga::NDTCell3d, beluga::detail::CellHasher<3>>>;
+
+/// Type of a particle-dependent random state generator.
+using RandomStateGenerator = std::function<typename beluga::NDTSensorModel<NDTMapRepresentation>::state_type()>;
```

Comment:
@serraramiro1 nit: consider aliasing `typename beluga::NDTSensorModel<NDTMapRepresentation>::state_type`, we use it more than once.